### PR TITLE
[Merged by Bors] - Feature: pod builder affinity and node selection

### PR DESCRIPTION
**Pull request by fhennig**
## Description

part of this ticket: https://github.com/stackabletech/issues/issues/300

Renamed from: https://github.com/stackabletech/operator-rs/tree/feat/affinity_0.25

Description: 
- Adds new PodBuilder methods for node selection and pod affinity (Also with tests)

## Review Checklist
- [x] Code contains useful comments
- ~~(Integration-)Test cases added (or not applicable)~~
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: src/builder/pod/mod.rs [hunk 1]
```diff
@@ -87,11 +86,31 @@ impl PodBuilder {
...
 
+    pub fn pod_anti_affinity(&mut self, anti_affinity: PodAntiAffinity) -> &mut Self {
+        self.pod_anti_affinity = Some(anti_affinity);
+        self
+    }
+
+    pub fn pod_affinity_opt(&mut self, affinity: Option<PodAffinity>) -> &mut Self {
+        self.pod_affinity = affinity;
+        self
+    }
```
Comment (maltesander) [inline src/builder/pod/mod.rs:97]: Can we move this below `pod_affinity` ?

File: src/builder/pod/mod.rs [hunk 2]
```diff
@@ -358,11 +377,12 @@ impl PodBuilder {
...
-                    ..Affinity::default()
+                    pod_anti_affinity: self.pod_anti_affinity.clone(),
                 })
                 .or_else(|| {
                     Some(Affinity {
                         pod_affinity: self.pod_affinity.clone(),
+                        pod_anti_affinity: self.pod_anti_affinity.clone(),
                         ..Affinity::default()
                     })
                 }),
```
Comment (maltesander) [inline src/builder/pod/mod.rs:388]: Couldnt this just be (maybe im missing something):
```
            affinity: Some(Affinity {
                node_affinity,
                pod_affinity: self.pod_affinity.clone(),
                pod_anti_affinity: self.pod_anti_affinity.clone(),
            }),
```

File: src/builder/pod/mod.rs [hunk 3]
```diff
@@ -406,33 +426,71 @@ impl PodBuilder {
...
+    #[fixture]
+    fn dummy_container() -> Container {
+        ContainerBuilder::new("container")
             .expect("ContainerBuilder not created")
-            .image("stackable/zookeeper:2.4.14")
-            .command(vec!["zk-server-start.sh".to_string()])
-            .args(vec!["stackable/conf/zk.properties".to_string()])
-            .add_volume_mount("zk-worker-1", "conf/")
-            .build();
+            .image("private-comapany/product:2.4.14")
```
Comment (maltesander) [inline src/builder/pod/mod.rs:445]:
```suggestion
            .image("private-company/product:2.4.14")
```

File: src/builder/pod/mod.rs [hunk 4]
```diff
@@ -446,7 +504,34 @@ mod tests {
...
+            .args(vec!["stackable/conf/zk.properties".to_string()])
+            .add_volume_mount("zk-worker-1", "conf/")
+            .build();
+
+        let init_container = ContainerBuilder::new("init-containername")
+            .expect("ContainerBuilder not created")
+            .image("stackable/zookeeper:2.4.14")
+            .command(vec!["wrapper.sh".to_string()])
+            .args(vec!["12345".to_string()])
+            .build();
```
Comment (maltesander) [inline src/builder/pod/mod.rs:534]: I think only container name and image are required for a dummy test.

**Review by maltesander — CHANGES_REQUESTED** [review]
Some minor things

File: src/builder/pod/mod.rs (hunk 4 above)
Comment (fhennig) [reply src/builder/pod/mod.rs:534]: that code was already there, I didn't add it (The diff is wrong). But I can still make those changes if you want to

File: src/builder/pod/mod.rs (hunk 2 above)
Comment (fhennig) [reply src/builder/pod/mod.rs]: Ah yes! I didn't event see that. Fixed cef7ac7723e4019af0655769e46c881aaf1a86d8

File: src/builder/pod/mod.rs (hunk 1 above)
Comment (fhennig) [reply src/builder/pod/mod.rs]: Yes. 8f4eef3

File: src/builder/pod/mod.rs (hunk 4 above)
Comment (fhennig) [reply src/builder/pod/mod.rs]: I cleaned it up a bit: 9af360a

File: src/builder/pod/mod.rs (hunk 2 above)
Comment (fhennig) [reply src/builder/pod/mod.rs]: I missed the simplied property creation, I did that here: 8922df6

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by fhennig** [comment]
bors merge

**Pull request closed without merging**